src/pudl/io_managers.py

```diff
@@ -0,0 +1,354 @@
+"""Dagster IO Managers."""
+import os
+from pathlib import Path
+from sqlite3 import sqlite_version
+
+import pandas as pd
+import sqlalchemy as sa
+from dagster import Field, InputContext, IOManager, OutputContext, io_manager
+from packaging import version
+from sqlalchemy.exc import SQLAlchemyError
+
+import pudl
+from pudl.metadata.classes import Package
+
+logger = pudl.logging_helpers.get_logger(__name__)
+
+MINIMUM_SQLITE_VERSION = "3.32.0"
+
+
+class ForeignKeyError(SQLAlchemyError):
+    """Raised when data in a database violates a foreign key constraint."""
+
+    def __init__(
+        self, child_table: str, parent_table: str, foreign_key: str, rowids: list[int]
+    ):
+        """Initialize a new ForeignKeyError object."""
+        self.child_table = child_table
+        self.parent_table = parent_table
+        self.foreign_key = foreign_key
+        self.rowids = rowids
+
+    def __str__(self):
+        """Create string representation of ForeignKeyError object."""
+        return f"Foreign key error for table: {self.child_table} -- {self.parent_table} {self.foreign_key} -- on for rows {self.rowids}\n"
```

Comment:
Little grammar nit but this should read either "for rows" or "on rows" instead of "on for rows"?

Reply:
oopsies thank you!